File: BankTransferMvpApp/BankTransferMvpApp/Utils/ServiceLocator.m
#import "ServiceLocator.h"
#import "AccountTransferPresenter.h"
#import "PresenterTypeEnum.h"
#import "LocalAccountRepository.h"
#import "RemoteAccountRepository.h"
#import "AppLaunchPresenter.h"
#import "AdManagerLocationRepository.h"
#import "MapLocationRepository.h"


@implementation ServiceLocator {

}

+ (id)register:(PresenterTypeEnum)presenterType {
    switch (presenterType) {
        case AppLaunch:
            return [self generateAppLaunchPresenter];
        case AccountTransfer:
            return [self generateAccountTransferPresenter];
    }
    return nil;
}

+ (AccountTransferPresenter *)generateAccountTransferPresenter {
    id<IRemoteAccountRepository> remoteAccountRepository = [[RemoteAccountRepository alloc] init];
    id<ILocalAccountRepository> localAccountRepository = [[LocalAccountRepository alloc] init];
    return [[AccountTransferPresenter alloc] initWithRemote:remoteAccountRepository
                                                                                                 andLocal:localAccountRepository];
}

+ (AppLaunchPresenter *)generateAppLaunchPresenter {
    id<IMapLocationRepository> mapLocationRepository = [[MapLocationRepository alloc] init];
    id<IAdManagerRepository> adManagerRepository = [[AdManagerLocationRepository alloc] init];

    return [[AppLaunchPresenter alloc] initWithMap:mapLocationRepository andAdManager:adManagerRepository];
}

@end

File: AccountTransferMvpApp/UnitTests/src/com/gaddzeit/utils/ServiceLocator.java
package com.gaddzeit.utils;

import com.gaddzeit.presenters.AccountTransferPresenter;
import com.gaddzeit.presenters.IPresenter;
import com.gaddzeit.repositories.ILocalAccountRepository;
import com.gaddzeit.repositories.IRemoteAccountRepository;
import com.gaddzeit.repositories.LocalAccountRepository;
import com.gaddzeit.repositories.RemoteAccountRepository;

public class ServiceLocator {
    public static IPresenter resolve(PresenterTypeEnum presenterTypeEnum) {
        switch(presenterTypeEnum) {
            case AccountTransfer:
                return createNewAccountTransferPresenter();
            default:
                return null;
        }
    }

    private static AccountTransferPresenter createNewAccountTransferPresenter() {
        IRemoteAccountRepository remoteAccountRepository = new RemoteAccountRepository();
        ILocalAccountRepository localAccountRepository = new LocalAccountRepository();
        return new AccountTransferPresenter(remoteAccountRepository, localAccountRepository);
    }
}
